userspace/falco/app_actions/load_rules_files.cpp

```diff
@@ -93,25 +93,40 @@ application::run_result application::load_rules_files()
 		falco_configuration::read_rules_file_directory(path, m_state->config->m_loaded_rules_filenames, m_state->config->m_loaded_rules_folders);
 	}
 
-	for (const auto& filename : m_state->config->m_loaded_rules_filenames)
+	std::vector<std::string> rules_contents;
+	falco::load_result::rules_contents_t rc;
+
+	try {
+		read_files(m_state->config->m_loaded_rules_filenames.begin(),
+			   m_state->config->m_loaded_rules_filenames.end(),
+			   rules_contents,
+			   rc);
+	}
+	catch(falco_exception& e)
+	{
+		return run_result::fatal(e.what());
+	}
+
+	for(auto &filename : m_state->config->m_loaded_rules_filenames)
 	{
 		falco_logger::log(LOG_INFO, "Loading rules from file " + filename + "\n");
 		std::unique_ptr<falco::load_result> res;
 
-		res = m_state->engine->load_rules_file(filename);
+		res = m_state->engine->load_rules(rc.at(filename), filename);
 
-		if((!res->successful() || (m_options.verbose && res->has_warnings())))
+		if(!res->successful())
 		{
-			printf("%s\n",
-			       (m_state->config->m_json_output ?
-				res->as_json().dump().c_str() :
-				res->as_string(true).c_str()));
+			// Return the summary version as the error
+			return run_result::fatal(res->as_string(true, rc));
```

Comment:
Do you think it would make sense to use `m_state->config->m_json_output` here too? Here we have forked behavior: if we have an error, throw a string exception, if we only have warnings, print them as string or json depending on `m_state->config->m_json_output`.

I have no strong opinion here, but I was wondering if this makes sense to you and what's your thinking.

Reply:
I think json output is pretty much just for validation use cases (that is, when run with -V). In those cases we want stdout to contain the validation result as json so a downstream program can capture it. And that's what happens:

```
./userspace/falco/falco -V rules_warning.yaml -V rules_error.yaml -o json_output=true -v  2> >(sed 's/^/err: /g') > >(sed 's/^/out: /g')
err: Thu Aug 11 08:41:33 2022: Falco version: 0.0.0
err: Thu Aug 11 08:41:33 2022: Falco initialized with configuration file: /etc/falco/falco.yaml
err: Thu Aug 11 08:41:33 2022: Falco compiled for: x86_64
err: Thu Aug 11 08:41:33 2022: Enabled sources: syscall
err: Thu Aug 11 08:41:33 2022: Validating rules file(s):
err: Thu Aug 11 08:41:33 2022:    rules_warning.yaml
err: Thu Aug 11 08:41:33 2022:    rules_error.yaml
err: Error:
out: {"falco_load_results":[{"errors":[],"name":"rules_warning.yaml","successful":true,"warnings":[{"code":"LOAD_NO_EVTTYPE","codedesc":"A rule condition matches too many evt.type values. This has a significant performance penalty. Make the condition more specific by adding an evt.type field or further restricting the number of evt.type values in the condition.","context":{"locations":[{"item_name":"no_evttype","item_type":"rule","position":{"column":2,"filename":"rules_warning.yaml","line":16,"offset":594}}],"snippet":"- rule: no_evttype\n  ^\n"},"message":"Rule matches too many evt.type values. This has a significant performance penalty."}]},{"errors":[{"code":"LOAD_ERR_YAML_VALIDATE","codedesc":"Error validating internal structure of YAML file","context":{"locations":[{"item_name":"bad_macro","item_type":"macro","position":{"column":2,"filename":"rules_error.yaml","line":0,"offset":2}}],"snippet":"- macro: bad_macro\n  ^\n"},"message":"Item has no mapping for key 'condition'"}],"name":"rules_error.yaml","successful":false,"warnings":[]}]}
```

I think I'll change the application actions so *only* validate's stdout is changed by `-o json_output=true`. Does that seem reasonable?